src/text/raw_text_reader.rs

```diff
@@ -670,103 +222,56 @@ impl<T: ToIonDataSource> IonReader for RawTextReader<T> {
         Self: Sized,
         F: FnOnce(&[u8]) -> U,
     {
-        match self.current_value.as_ref().map(|current| current.value()) {
-            Some(TextValue::Clob(ref value)) => Ok(f(value.as_slice())),
-            _ => Err(self.expected("clob value")),
-        }
+        self.reader.map_clob(f)
     }
 
     fn read_timestamp(&mut self) -> IonResult<Timestamp> {
-        match self.current_value.as_ref().map(|current| current.value()) {
-            Some(TextValue::Timestamp(ref value)) => Ok(value.clone()),
-            _ => Err(self.expected("timestamp value")),
-        }
+        self.reader.read_timestamp()
     }
 
     fn step_in(&mut self) -> IonResult<()> {
-        match &self.current_value {
-            Some(value) if value.ion_type().is_container() => {
-                self.parents
-                    .push(ParentContainer::new(value.value().ion_type()));
-                self.current_value = None;
-                Ok(())
-            }
-            Some(value) => {
-                illegal_operation(format!("Cannot step_in() to a {:?}", value.ion_type()))
-            }
-            None => illegal_operation(format!(
-                "{} {}",
-                "Cannot `step_in`: the reader is not positioned on a value.",
-                "Try calling `next()` to advance first."
-            )),
-        }
+        self.reader.step_in()
     }
 
     fn step_out(&mut self) -> IonResult<()> {
-        if self.parents.is_empty() {
-            return illegal_operation(
-                "Cannot call `step_out()` when the reader is at the top level.",
-            );
-        }
-
-        // The container we're stepping out of.
-        let parent = self.parents.last().unwrap();
-
-        // If we're not at the end of the current container, advance the cursor until we are.
-        // Unlike the binary reader, which can skip-scan, the text reader must visit every value
-        // between its current position and the end of the container.
-        if !parent.is_exhausted() {
-            while let RawStreamItem::Value(_) | RawStreamItem::Null(_) = self.next()? {}
-        }
-
-        // Remove the parent container from the stack and clear the current value.
-        let _ = self.parents.pop();
-        self.current_value = None;
-
-        if self.parents.is_empty() {
-            // We're at the top level; nothing left to do.
-            return Ok(());
-        }
+        self.feed_if_needed()?;
```

Comment:
I suggest making a `retry_on_incomplete` helper method that takes a closure to loop over as long as it's returning `Incomplete`. Then we can use it in `step_in()`, `step_out()` and (when this wrapper is format-agnostic) all of the `read_*` calls that can trigger I/O in the binary reader.

Reply:
With the recent commit, the retry logic is only contained in the `next`, and `step_out` methods. I do want to pull it into a helper method, but ran into issues due to borrows of self. I need to spend a little more time with it, but didn't want to hold up the PR any more than it already has been.